File: demo-material3/lib/features/peer_evaluation/data/repositories/assessment_repository.dart
import '../../domain/models/assessment.dart';
import '../../domain/models/course.dart';
import '../../domain/models/criteria.dart';
import '../../domain/models/member_result.dart';
import '../../domain/models/peer.dart';
import '../../domain/repositories/i_assessment_repository.dart';
import '../datasources/i_assessment_source.dart';

class AssessmentRepository implements IAssessmentRepository {
  final IAssessmentSource source;

  AssessmentRepository({required this.source});

  @override
  Future<List<Assessment>> getAssessments() async =>
      source.getAssessments();

  @override
  Future<List<Criteria>> getCriteria() async => source.getCriteria();

  @override
  Future<List<Course>> getCourses() async => source.getCourses();

  @override
  Future<List<Peer>> getPeers() async => source.getPeers();

  @override
  Future<List<MemberResult>> getMemberResults() async {
    final peers = source.getPeers();

    // Fixed scores per peer id matching iOS evaluation data for assess-1 peers
    const Map<String, Map<String, double>> iosScores = {
      'peer-1': {'Puntualidad': 4.5, 'Contribuciones': 3.5, 'Compromiso': 3.5, 'Actitud': 4.0},
      'peer-2': {'Puntualidad': 4.0, 'Contribuciones': 4.7, 'Compromiso': 4.3, 'Actitud': 4.7},
      'peer-3': {'Puntualidad': 3.5, 'Contribuciones': 4.0, 'Compromiso': 3.8, 'Actitud': 3.5},
      'peer-4': {'Puntualidad': 4.2, 'Contribuciones': 3.8, 'Compromiso': 4.0, 'Actitud': 4.5},
    };

    // Only include peers that have fixed iOS scores (assess-1 peers)
    return peers
        .where((peer) => iosScores.containsKey(peer.id))
        .map((peer) => MemberResult(peer: peer, scores: iosScores[peer.id]!))
        .toList();
  }
}

File: demo-liquid-glass/DomainKit/Sources/DomainKit/data/repositories/AssessmentRepository.swift
// data/repositories/AssessmentRepository.swift
// Concrete implementation of AssessmentRepositoryProtocol
// Depends on LocalAssessmentSource (data layer) — never on presentation

public final class AssessmentRepository: AssessmentRepositoryProtocol {
    private let source: LocalAssessmentSource

    public init(source: LocalAssessmentSource) {
        self.source = source
    }

    public func getCourses() -> [Course] {
        source.fetchCourses()
    }

    public func getAssessments() -> [Assessment] {
        source.fetchAssessments()
    }

    public func getCriteria() -> [Criteria] {
        source.fetchCriteria()
    }

    public func getPeers() -> [Peer] {
        source.fetchPeers()
    }

    public func getEvaluations(for assessmentId: String) -> [Evaluation] {
        source.fetchEvaluations(for: assessmentId)
    }

    /// Aggregates all evaluations for an assessment into per-peer MemberResult values.
    public func computeResults(for assessmentId: String) -> [MemberResult] {
        let evaluations = source.fetchEvaluations(for: assessmentId)
        let allPeers = source.fetchPeers()
        let criteria = source.fetchCriteria()

        // Group evaluations by peerId
        let grouped = Dictionary(grouping: evaluations, by: { $0.peerId })

        return grouped.compactMap { (peerId, peerEvaluations) -> MemberResult? in
            guard let peer = allPeers.first(where: { $0.id == peerId }) else { return nil }

            // Compute average per criteria across all evaluators
            var criteriaAverages: [String: Double] = [:]
            for crit in criteria {
                let critScores = peerEvaluations.compactMap { eval -> Double? in
                    guard let score = eval.scores[crit.id] else { return nil }
                    return Double(score)
                }
                if !critScores.isEmpty {
                    criteriaAverages[crit.name] = critScores.reduce(0, +) / Double(critScores.count)
                }
            }

            return MemberResult(peer: peer, scores: criteriaAverages)
        }
        .sorted { $0.average > $1.average }
    }
}
